test/e2e/0xTrade.test.ts

```diff
@@ -0,0 +1,297 @@
+  };
+
+  beforeEach(async () => {
+    const deployment = await deployTestContracts();
+
+    ({
+      deployer,
+      settlement,
+      allowanceManager,
+      wallets: [solver, trader, marketMaker],
+    } = deployment);
+
+    const { authenticator, owner } = deployment;
+    await authenticator.connect(owner).addSolver(solver.address);
+
+    const { chainId } = await ethers.provider.getNetwork();
+    domainSeparator = domain(chainId, settlement.address);
+
+    owl = await waffle.deployContract(deployer, ERC20, ["OWL", 18]);
+    gno = await waffle.deployContract(deployer, ERC20, ["GNO", 18]);
+
+    const exchange = await waffle.deployContract(
+      deployer,
+      Exchange.compilerOutput,
+      [chainId],
+    );
+    const erc20Proxy = await waffle.deployContract(
+      deployer,
+      ERC20Proxy.compilerOutput,
+    );
+
+    await erc20Proxy.addAuthorizedAddress(exchange.address);
+    await exchange.registerAssetProxy(erc20Proxy.address);
+
+    zeroEx = {
+      exchange,
+      erc20Proxy,
+      // NOTE: Domain separator parameters taken from:
+      // <https://0x.org/docs/guides/v3-specification#eip-712-usage>
+      domainSeparator: {
+        name: "0x Protocol",
+        version: "3.0.0",
+        chainId,
+        verifyingContract: exchange.address,
+      },
+    };
+  });
+
+  it("should settle an EOA trade with a 0x trade", async () => {
+    // Settles a market order buying 1 GNO for 120 OWL and get matched with a
+    // market maker using 0x orders.
+
+    await owl.mint(trader.address, ethers.utils.parseEther("140"));
+    await owl
+      .connect(trader)
+      .approve(allowanceManager.address, ethers.constants.MaxUint256);
+
+    await gno.mint(marketMaker.address, ethers.utils.parseEther("1000.0"));
+    await gno
+      .connect(marketMaker)
+      .approve(zeroEx.erc20Proxy.address, ethers.constants.MaxUint256);
+
+    const gpv2Order = {
+      kind: OrderKind.BUY,
+      partiallyFillable: false,
+      buyToken: gno.address,
+      sellToken: owl.address,
+      buyAmount: ethers.utils.parseEther("1.0"),
+      sellAmount: ethers.utils.parseEther("130.0"),
+      feeAmount: ethers.utils.parseEther("10.0"),
+      validTo: 0xffffffff,
+      appData: 1,
+    };
+
+    const zeroExGnoPrice = 110;
+    const zeroExSignedOrder = await signZeroExSimpleOrder(
+      marketMaker,
+      zeroEx.domainSeparator,
+      {
+        makerAssetAddress: gno.address,
+        makerAssetAmount: ethers.utils.parseEther("1000.0"),
+        takerAssetAddress: owl.address,
+        takerAssetAmount: ethers.utils.parseEther("1000.0").mul(zeroExGnoPrice),
+      },
+    );
+    expect(
+      await zeroEx.exchange.isValidOrderSignature(
+        zeroExSignedOrder.order,
+        zeroExSignedOrder.signature,
+      ),
+    ).to.be.true;
+
+    const encoder = new SettlementEncoder(domainSeparator);
+    await encoder.signEncodeTrade(gpv2Order, trader, SigningScheme.TYPED_DATA);
+    encoder.encodeInteraction({
```

Comment:
If approval are pre-set, then we can have it even cheaper, right?

Reply:
I would expect so.

Reply:
Just did a quick and dirty test and its almost exactly 20gas cheaper:
```
  test:e2e:0xTrade gas used: 252854 +0ms
```

This is definitely something that the driver can optimize (by setting unlimited approval on the 0x contract once per token).